src/view/ApplicationView.tsx

```diff
@@ -0,0 +1,185 @@
+): Component is FCLike<Props> {
+  return !(Component.prototype && Component.prototype.isReactComponent);
+}
+
+/**
+ * All components that reference observable stores, either directly or through props, should be
+ * wrapped in View() so that they rerender when the store fields they access are changed.
+ *
+ * Heavily adapted from https://github.com/mobxjs/mobx/blob/main/packages/mobx-react-lite/src/useObserver.ts
+ */
+export default function View<Props extends {}, TRef>(
+  Component:
+    | React.ComponentClass<Props>
+    | FCLike<Props>
+): React.FunctionComponent<React.PropsWithRef<Props>> {
+  if (isFunctionComponent(Component)) {
+    const ReactiveComponent = React.forwardRef((props: Props, ref) => {
+      const viewStoreRef = useRef<ViewStore<typeof Component> | null>(null);
+      if (viewStoreRef.current == null) {
+        // ok to write ref during render since this is idempotent
+        viewStoreRef.current = new ViewStore(Component);
+        // renders can be abandoned & components can be remounted in StrictMode/Concurrent Mode, without any
+        // effects firing to tell us when a ViewStore is no longer being used. A FinalizationRegistry
+        // lets us _eventually_ catch when an old ref is disposed, and run some cleanup.
+        // 2nd arg is value passed to FinalizationRegistry callback, 3rd arg lets us unregister
+        reactionRegistry.register(
+          viewStoreRef,
+          viewStoreRef.current,
+          viewStoreRef.current
+        );
+      }
+      const viewStore = viewStoreRef.current;
+
+      // force component to rerender when relevant fields of store change
+      const _dummy = useSyncExternalStore(
+        viewStore.subscribe,
+        viewStore.getSnapshot,
+        // MobX includes the server snapshot, we may not need it but doesn't hurt
+        viewStore.getSnapshot
+      );
+
+      // run the reactive render instead of the original one, idempotent
+      const render = viewStore.getReaction();
+      return render(props, ref);
+    });
+    ReactiveComponent.displayName = `View(${Component.displayName ?? Component.name ?? 'Component'})`;
+    // TODO: Fix or ignore this type error
+    return React.memo(ReactiveComponent);
+  }
+
+  // Same thing, except for class components, though we don't seem to be wrapping any class
+  // components in aha-app. Doesn't prevent tearing, since class components can't use
+  // `useSyncExternalStore`. The original plan was to render the class component under a function
+  // component similar to the above, only observing the class component's render method, but there's
+  // no way to reference the render method from the function component before rendering is complete.
+  // This is the best we can do.
+  // TODO: Actually, maybe we can pass a ref from the parent function component to the child class
+  // component and somehow use that to wire up the ViewStore with the class's `render()` method? Idk
+  // if that's possible.
```

Comment:
I'm not sure if we need to support class components? The ones we have would date to pre-mvc I think, and if we were to adopt mvc we'd migrate them to function components.

Reply:
I agree—I just hate to do nothing and risk silent failures should we start wrapping class components. But maybe TS types that reject class components and documentation is enough.